repos/system_upgrade/el7toel8/actors/checkkdegnome/actor.py

```diff
@@ -0,0 +1,137 @@
+    """
+    name = 'check_kde_gnome'
+    consumes = ()
+    produces = (CanUpgrade,)
+    tags = (RhelUpgradeTag,)
+
+    def is_executable(self, path):
+        """
+        Checks if path exists, if it is file and if is executable.
+        """
+        return os.path.exists(path) and os.path.isfile(path) and os.access(path, os.X_OK)
+
+    def get_xsession(self, path):
+        """
+        Gets current XSession. Return empty string if no XSession found (any reason
+        including bad path etc.)
+        """
+        default_xsession = ""
+        if not (os.path.exists(path) and os.path.isfile(path)):  # Bad path - in container for example
+            return default_xsession
+        with open(path, "r") as f:
+            for line in f.readlines():
+                if "xsession" in line.lower():
+                    default_xsession = line.split("=")[1].lower()
+        return default_xsession
+
+    def check_app_in_use(self, app):
+        """
+        Method return True if application was used in last month, False in other cases.
+        """
+        path = "{0}/.kde/share/config/{1}rc".format(os.environ.get("HOME"), app)
+        if os.path.exists(path) and os.path.isfile(path):
+            last_modified = os.stat(path).st_mtime
+            if last_modified >= int(time.time() - 2592000):
+                self.log.info("Application {0} is actively used".format(app))
+                return True
+        return False
+
+    def is_installed(self, app):
+        """
+        Wrapper for "rpm -q <app>" command, returns its return value, output is not supressed
+        """
+        return subprocess.call(["rpm", "-q", app])
+
+    def process(self):
+        apps_in_use = 0
+        self.log.info("  Detecting desktop environments  ")
+        self.log.info("==================================")
+
+        # Detect installed desktops by their startup files
+        kde_desktop_installed = self.is_executable("/usr/bin/startkde")
+        gnome_desktop_installed = self.is_executable("/usr/bin/gnome-session")
+        self.log.info("* KDE installed: {0}".format(kde_desktop_installed))
+        self.log.info("* Gnome installed: {0}".format(gnome_desktop_installed))
+        self.log.info("----------------------------------")
+
+        # No desktop installed, we don't even care about apps as they are most likely not used or even installed
+        if not kde_desktop_installed and not gnome_desktop_installed:
+            self.log.info("INFO: No desktop installed. Continuing with the upgrade.")
+            self.produce(CanUpgrade(canUpgrade=0))
+            return
+
+        if kde_desktop_installed:
+            self.log.info("INFO: KDE desktop is installed. Checking what we can do about it.")
+            if not gnome_desktop_installed:
+                self.log.info("ERROR: Cannot perform the upgrade because there is no other desktop installed.")
+                self.produce(CanUpgrade(canUpgrade=1))
```

Comment:
Please do not create a message which nobody consumes. You can remove canUpgrade message model with the topic RhelUpgradeInfoTopic as well.
Instead of that, use `create_report` function provided by the leapp framework. To not allow the upgrade, use `reporting.Flags.INHIBITOR`. See example here: https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/actors/checkfirewalld/actor.py#L35.